.github/workflows/rosetta-api.yml

```
@@ -46,34 +46,77 @@ jobs:
           path: ./**/*.tgz
           if-no-files-found: error
 
-  validate:
+  image:
     if: github.event_name == 'push' || (github.event.pull_request.head.repo.full_name == github.repository && github.actor != 'dependabot[bot]')
     runs-on: ubuntu-latest
-    timeout-minutes: 20
+    timeout-minutes: 10
     steps:
       - name: Checkout
         uses: actions/checkout@v2
 
-      - name: Build Mirror Node All-in-One
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Set GIT_REF
         run: |
           GIT_REF=${GITHUB_HEAD_REF:-$GITHUB_REF}
-          docker build --build-arg GIT_REF=${GIT_REF#refs/*/} --tag="${MODULE}" .
-        working-directory: ./hedera-mirror-rosetta/build
+          GIT_REF=${GIT_REF#refs/*/}
+          echo "GIT_REF=${GIT_REF}" >> $GITHUB_ENV
+
+      - name: Build Mirror Node All-in-One Docker Image and Export
+        uses: docker/build-push-action@v2
+        with:
+          build-args: GIT_REF=${{ env.GIT_REF }}
+          context: ./hedera-mirror-rosetta/build
+          tags: ${{ env.MODULE }}:latest
+          outputs: type=docker,dest=/tmp/${{ env.MODULE }}.tar
+
+      - name: Upload Artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ env.MODULE }}
+          path: /tmp/${{ env.MODULE }}.tar
+          if-no-files-found: error
+
+  validate:
+    runs-on: ubuntu-latest
+    needs: image
+    strategy:
+      matrix:
+        online-only: ['true', 'false']
```

```suggestion
        mode: ['online', 'offline']
```

Then you can pass this to docker run with `-e MODE=${{ matrix.mode }}`

---

rosetta-cli validation requires at least the online rosetta server. rosetta-cli validation also support interacting with an additional offline mode rosetta server mainly for Construction API endpoints which do not require online interactions.

so the testing matrix is correct:

- online-only 'true', there is a single online mode rosetta server, and we use the online server URL as the `offline_url` for rosetta-cli validation
- online-only 'false', one online mode rosetta server and one offline mode rosetta server, the urls for rosetta-cli are configured correspondently

if we only spin up an offline mode rosetta server, there is no way to validate either DATA or Construction APIs.

---

I still don't think this is an appropriate approach. Tests should model how the server will be used in production. No one will be running both an online and offline server simultaneously in production. The rosetta-cli supports testing just construction, just data, or both. Why can't we always run the construction tests and conditionally run data if online?

---

agree that we should model what's being used in production. and in fact, a paired online mode and offline mode servers are closer to what exchanges will use in production. According to [rosetta spec](https://www.rosetta-api.org/docs/construction_api_introduction.html#flow-of-operations), for construction API, only `/construction/metadata` and `/construction/submit` are marked with `online`.

on the other hand, we also need to test with online only mode so as to make sure the APIs which are supposed to work in both modes meet the expectations.

---

Ok, I wish the rosetta-cli could test the online and offline modes in isolation but seems that's not the case. So I'm fine with the approach.